analyzers/src/SonarAnalyzer.CSharp/Rules/NonDerivedPrivateClassesShouldBeSealed.cs

```diff
@@ -24,48 +24,78 @@ namespace SonarAnalyzer.Rules.CSharp
     public sealed class NonDerivedPrivateClassesShouldBeSealed : SonarDiagnosticAnalyzer
     {
         private const string DiagnosticId = "S3260";
-        private const string MessageFormat = "Private classes or records which are not derived in the current assembly should be marked as 'sealed'.";
+        private const string MessageFormat = "{0} {1} which are not derived in the current assembly should be marked as 'sealed'.";
 
         private static readonly DiagnosticDescriptor Rule = DescriptorFactory.Create(DiagnosticId, MessageFormat);
-
         public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray.Create(Rule);
 
+        private static readonly ImmutableHashSet<SyntaxKind> KindsToBeDescended = ImmutableHashSet.Create(
+            SyntaxKind.CompilationUnit,
+            SyntaxKind.NamespaceDeclaration,
+            SyntaxKindEx.FileScopedNamespaceDeclaration,
+            SyntaxKind.InterfaceDeclaration,
+            SyntaxKind.ClassDeclaration,
+            SyntaxKind.StructDeclaration,
+            SyntaxKindEx.RecordClassDeclaration,
+            SyntaxKindEx.RecordStructDeclaration);
+
+        private static readonly ImmutableHashSet<SyntaxKind> PossiblyVirtualKinds = ImmutableHashSet.Create(
+            SyntaxKind.MethodDeclaration,
+            SyntaxKind.PropertyDeclaration,
+            SyntaxKind.EventDeclaration,
+            SyntaxKind.IndexerDeclaration);
+
         protected override void Initialize(SonarAnalysisContext context) =>
-            context.RegisterNodeAction(c =>
+            context.RegisterTreeAction(c =>
             {
-                var typeDeclarationSyntax = (TypeDeclarationSyntax)c.Node;
-                if (!c.IsRedundantPositionalRecordContext()
-                    && IsPrivateButNotSealedType(typeDeclarationSyntax)
-                    && !HasVirtualMembers(typeDeclarationSyntax))
+                var declarations = c.Tree.GetCompilationUnitRoot().DescendantNodes(x => x.IsAnyKind(KindsToBeDescended))
```

Comment:
Educational: what is the difference between `.GetCompilationUnitRoot()` and `.GetRoot()`?

Reply:
`GetCompilationUnitRoot` calls `GetRoot` and casts it to `CompilationUnitSyntax`.
Technically, all I care about is base `SyntaxNode` functionality, so I will change it to `GetRoot` to avoid a useless cast.